ui/dashboards/src/utils/time-range-params.ts

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useMemo, useEffect, useRef } from 'react';
+import { useQueryParams, DecodedValueMap, StringParam } from 'use-query-params';
+import { getUnixTime, sub } from 'date-fns';
+import {
+  TimeRangeValue,
+  AbsoluteTimeRange,
+  toAbsoluteTimeRange,
+  isRelativeTimeRange,
+  isDurationString,
+  RelativeTimeRange,
+  DurationString,
+  isAbsoluteTimeRange,
+  parseDurationString,
+} from '@perses-dev/core';
+import { useSelectedTimeRange } from '../context';
+
+const timeRangeQueryConfig = {
+  start: StringParam,
+  end: StringParam,
+};
+
+/**
+ * Gets the initial time range taking into account URL params and dashboard JSON duration
+ * Sets start query param if it is empty on page load
+ */
+export function useInitialTimeRange(dashboardDuration: DurationString): TimeRangeValue {
+  const [query, setQuery] = useQueryParams(timeRangeQueryConfig);
+  const { start, end } = query;
+
+  return useMemo(() => {
+    let initialTimeRange: TimeRangeValue = { pastDuration: dashboardDuration };
+    if (!start) {
+      setQuery({ start: dashboardDuration });
+      return initialTimeRange;
+    }
+
+    if (isDurationString(start.toString())) {
+      initialTimeRange = { pastDuration: start } as RelativeTimeRange;
+    } else {
+      initialTimeRange = {
+        start: new Date(Number(start)),
+        end: end ? new Date(Number(end)) : new Date(),
+      };
+    }
+    return initialTimeRange;
+  }, [start, end, dashboardDuration, setQuery]);
+}
+
+/**
+ * Set start and end URL params and update when selected time range changes
+ */
+export function useSyncTimeRangeParams() {
```

Comment:
Wondering if there is a case for when we don't want URL params, so maybe a boolean `enabled` could be nice too. For example, if we have two instances of the DashboardView that need different time ranges

```tsx
useSyncTimeRangeParams(enabled: boolean)
```

Reply:
Yeah good point, similar to how it was possible to "disable" them before by leaving setQueryString undefined, I'll add this boolean!

Reply:
useSyncTimeRangeParams is gone now (see back btn thread), but I did add [this paramsEnabled boolean](https://github.com/perses/perses/blob/1b5377230d8a9cc04796019f4c084dc3e574a0ce/ui/dashboards/src/utils/time-range-params.ts#L119) to useSetTimeRangeParams!